### ISSUE TITLE
onCalculated event fires at unexpected times with custom functions

### ISSUE THREAD

**Issue by fds-ssg** [OfficeDev/office-js#587]
## Summary
There are actually a few issues:
1. The onCalculated event for a Worksheet or the WorksheetCollection is firing before the result is returned to a custom function.
1. For Excel online, the onCalculated event does not fire after the final formula result is delivered.
1. The behavior is different for Excel online than it is for PC Excel.

## Expected Behavior
1. The onCalculated event should not fire until results have been returned (promise either resolved or rejected) for a function being recalculated.
1. If a recalculation occurrs in a cell with dependents, onCalculated should fire after all functions in a chain of dependencies have had results returned.
1. The behavior for when onCalculated is fired should be identical for Excel on the PC and online.

## Current Behavior

The behavior described here occurs with the steps to reproduce below, which basically consist of the following setup:
* Task pane code that hooks the onCalculated event for both the active sheet and the WorksheetCollection
* Custom function that increments and returns a count, after a delay specified as a function argument
* Worksheet with two custom functions, one of which references the other

Excel online
- onCalculated fires before promise is returned from first custom function invocation
- onCalculated fires only once, even if there is a chain of references
- onCalculated does not fire after result returned from final function in the chain

Excel PC
- onCalculated fires before any value is returned from first custom function invocation
- onCalculated fires three times during evaluation of two formulas
- onCalculated does fire after result returned from final function in the chain

## Steps to Reproduce, or Live Example

* Additional details:
  1. Import the TestOnCalc gist into ScriptLab: https://gist.github.com/fds-ssg/00855a5ef0e7670f3ee5b76e6d87bee5
  2. Create and register the FDSTEST function
  3. Enter the following two formulas into a worksheet:
cell A1: =FDSTEST("inc",2500)
cell A2: =FDSTEST("inc",A1+2000)
  4. Click the "Hook worksheet collection" and "Hook active sheet" buttons on the TestOnCalc task pane
  5. Edit cell A1 so that it recalculates

*Custom function code*
```
  let globalCount = 0;
  async function fdstest(input: string, input2: string, invoke: CustomFunctions.Invocation) {
    console.log(`fdstest-impl ${invoke.address}: fdsTest("${input}","${input2}")`);
    let result: string = input;

    if (input === 'inc') {
      // Increment test; add to the global formula count and return the result
      result = (++globalCount).toString();
      if (input2) {
        const delay = parseInt(input2);
        return new Promise(resolve => {setTimeout(()=>resolve(result),delay);});
      }
    }
    return new Promise(resolve => resolve(result));
  }
```

*Custom function json*
```
    {
      "description": "Write a message to the log.",
      "helpUrl": "http://www.factset.com/",
      "id": "FDSTEST",
      "name": "FDSTEST",
      "parameters": [
        {
          "description": "The message",
          "name": "message",
          "type": "string",
          "dimensionality": "scalar"
        },
        {
          "description": "Optional parameter",
          "name": "message2",
          "optional": true,
          "type": "string",
          "dimensionality": "scalar"
        }
      ],
      "options": {
        "requiresAddress": true
      },
      "result": {
        "type": "string",
        "dimensionality": "scalar"
      }
    },
```

## Your Environment

* Platform [PC desktop, Mac, iOS, Office Online]: Excel desktop and online
* Host [Excel, Word, PowerPoint, etc.]: Excel
* Office version number:  Online - 16.0.11823.31875, PC - Version 1907 (Build 11819.20002)
* Operating System: Windows 10
* Browser (if using Office Online): Chrome

### Excel online
For Excel on the web, the onCalculated events fire prior to custom functions being evaluated, as demonstrated by the following log:
```
index.tsx:78 SH-Sheet1: Calculated completed!
index.tsx:78 SH-Sheet1: Event Args Detail: {"type":"WorksheetCalculated","worksheetId":"{00000000-0001-0000-0000-000000000000}"}
index.tsx:78 COL: Calculated completed!
index.tsx:78 COL: Event Args Detail: {"type":"WorksheetCalculated","worksheetId":"{00000000-0001-0000-0000-000000000000}"}
functions.ts:90 Evaluating Sheet1!A1: fdsTest("inc","2500")
functionsImpl.ts:271 fdstest-impl Sheet1!A1: fdsTest("inc","2500")
index.tsx:78 SH-Sheet1: Calculation completed on worksheet: Sheet1
index.tsx:78 COL: Calculation completed on worksheet: Sheet1
functions.ts:94   Result Sheet1!A1: fdsTest("inc","2500") = 14
functions.ts:90 Evaluating Sheet1!A2: fdsTest("inc","2014")
functionsImpl.ts:271 fdstest-impl Sheet1!A2: fdsTest("inc","2014")
functions.ts:94   Result Sheet1!A2: fdsTest("inc","2014") = 15
```
### Excel on the PC
For Excel on the PC, onCalculated fires multiple times.  As in the web scenario it fires prior to evaluation of the functions, but it fires again after each function has been evaluated.

Unfortunately it's hard to get a consolidated log on the PC for Custom Functions runtime code and Task Pane runtime code, so here are separate logs for both.

*Task Pane (Script Lab) log*
(with annotations indicating activity in main Excel window)
```
============  Edit cell A1, both A1 and A2 show #BUSY
SH-Sheet1: Calculated completed!
SH-Sheet1: Event Args Detail: {"type":"WorksheetCalculated","worksheetId":"{00000000-0001-0000-0000-000000000000}"}
COL: Calculated completed!
COL: Event Args Detail: {"type":"WorksheetCalculated","worksheetId":"{00000000-0001-0000-0000-000000000000}"}
SH-Sheet1: Calculation completed on worksheet: Sheet1
COL: Calculation completed on worksheet: Sheet1
=============   Result returned for A1, A2 still #BUSY
SH-Sheet1: Calculated completed!
SH-Sheet1: Event Args Detail: {"type":"WorksheetCalculated","worksheetId":"{00000000-0001-0000-0000-000000000000}"}
COL: Calculated completed!
COL: Event Args Detail: {"type":"WorksheetCalculated","worksheetId":"{00000000-0001-0000-0000-000000000000}"}
SH-Sheet1: Calculation completed on worksheet: Sheet1
COL: Calculation completed on worksheet: Sheet1
=============   Result returned for A2
SH-Sheet1: Calculated completed!
SH-Sheet1: Event Args Detail: {"type":"WorksheetCalculated","worksheetId":"{00000000-0001-0000-0000-000000000000}"}
COL: Calculated completed!
COL: Event Args Detail: {"type":"WorksheetCalculated","worksheetId":"{00000000-0001-0000-0000-000000000000}"}
SH-Sheet1: Calculation completed on worksheet: Sheet1
COL: Calculation completed on worksheet: Sheet1
```

*Function log*
```
6/28/2019 15:11:50	Verbose	CustomFunctions	[Execution] [Async] [Begin] Function=FDSTEST, Workbook=excel add-in 367F4E28-F27A-41A8-92A6-9F79D56CB19E.xlsx		
6/28/2019 15:11:50	Verbose	Runtime	[Console] [Log] Verbose CustomFunctions [Execution] [Begin] Function=FDSTEST		
6/28/2019 15:11:50	Verbose	Runtime	[Console] [Log] Evaluating Sheet1!A1: fdsTest("inc","2500")		
6/28/2019 15:11:50	Verbose	Runtime	[Console] [Log] fdstest-impl Sheet1!A1: fdsTest("inc","2500")		
6/28/2019 15:11:53	Verbose	Runtime	[Console] [Log]   Result Sheet1!A1: fdsTest("inc","2500") = 7		
6/28/2019 15:11:53	Verbose	Runtime	[Console] [Log] Medium CustomFunctions [Execution] [End] [Success] Function=FDSTEST		
6/28/2019 15:11:53	Medium	CustomFunctions	[Execution] [Async] [End] [Success] Function=FDSTEST, Workbook=excel add-in 367F4E28-F27A-41A8-92A6-9F79D56CB19E.xlsx		
6/28/2019 15:11:53	Verbose	CustomFunctions	[Execution] [Async] [Begin] Function=FDSTEST, Workbook=excel add-in 367F4E28-F27A-41A8-92A6-9F79D56CB19E.xlsx		
6/28/2019 15:11:53	Verbose	Runtime	[Console] [Log] Verbose CustomFunctions [Execution] [Begin] Function=FDSTEST		
6/28/2019 15:11:53	Verbose	Runtime	[Console] [Log] Evaluating Sheet1!A2: fdsTest("inc","2007")		
6/28/2019 15:11:53	Verbose	Runtime	[Console] [Log] fdstest-impl Sheet1!A2: fdsTest("inc","2007")		
6/28/2019 15:11:55	Verbose	Runtime	[Console] [Log]   Result Sheet1!A2: fdsTest("inc","2007") = 8		
6/28/2019 15:11:55	Verbose	Runtime	[Console] [Log] Medium CustomFunctions [Execution] [End] [Success] Function=FDSTEST		
6/28/2019 15:11:55	Medium	CustomFunctions	[Execution] [Async] [End] [Success] Function=FDSTEST, Workbook=excel add-in 367F4E28-F27A-41A8-92A6-9F79D56CB19E.xlsx		
```

**Comment by keyur32** [comment]
Internal VSO Bug #3494909

**Comment by keyur32** [comment]
Thanks for the issue. Team has been investigating and will currently looking at either only firing the calc ended event or providing a separate event for Calc Ended when it comes to custom functions.

For now, the recommendation would be to track any pending promises (using the batching pattern: https://docs.microsoft.com/en-us/office/dev/add-ins/excel/custom-functions-batching). If no pending promises remain from your function, then you can safely apply post processing as this means all functions have returned successfully.

**Comment by SWDevB** [comment]
I also can confirm the behavior on Excel Online, but the solution is a bit more complex than suggested, because as far as I know for example for nested functions, the outer function is called after the inner function was resolved and of course the inner function knows about being an inner function. So there is a change that right after resolving the "last" promise the next one is already incoming.

**Comment by weshi** [comment]
The fix to make the onCalculated event fired after the custom function calculation is finished is already available in Excel Online. For the nested functions above, it's a different issue, and it should be a by-design behavior. Please create a new issue this if you think differently. Thanks.

**Comment by didu9898** [comment]
The issue is still reproducible on Excel for the web. Although I don't have exact steps to reproduce it, the problem occurs with a custom function I use. I rely on the `onCalculated` event to determine when the calculation is complete and the text is rendered in the sheet. However, there's about a 1-second delay between when the `onCalculated` event is triggered and when the text actually appears in the sheet. I've also tried loading `range.text` and comparing it with the expected text to confirm if the text has rendered, but this approach hasn't been helpful. The delay is noticeable when cell still contains 'Busy' and the event is fired before the text is rendered.
